libecole/src/observation/milpbipartite.cpp

```diff
@@ -0,0 +1,311 @@
+	std::tuple<std::vector<scip::Var*>, std::vector<scip::real>, std::optional<scip::real>, std::optional<scip::real>>> {
+	SCIP_Bool success = false;
+	int n_constraint_variables;
+	int n_active_variables;
+	scip::real constant_offset = 0;
+	int requiredsize = 0;
+
+	// Find how many active variables and constraint variables there are (for allocation)
+	scip::call(SCIPgetConsNVars, scip, constraint, &n_constraint_variables, &success);
+	if (!success) return {};
+	n_active_variables = SCIPgetNVars(scip);
+
+	// Allocate buffers large enough to hold future variables and coefficients
+	int buffer_size = std::max(n_constraint_variables, n_active_variables);
+	auto* variables = new scip::Var*[static_cast<std::size_t>(buffer_size)];
+	auto* coefficients = new scip::real[static_cast<std::size_t>(buffer_size)];
+
+	// Get the variables and their coefficients in the constraint
+	scip::call(SCIPgetConsVars, scip, constraint, variables, buffer_size, &success);
+	if (!success) return {};
+	scip::call(SCIPgetConsVals, scip, constraint, coefficients, buffer_size, &success);
+	if (!success) return {};
+
+	// If we are in SCIP_STAGE_TRANSFORMED or later, the variables in the constraint might be inactive
+	// Re-express the coefficients in terms of active variables
+	if (SCIPgetStage(scip) >= SCIP_STAGE_TRANSFORMED) {
+		scip::call(
+			SCIPgetProbvarLinearSum,
+			scip,
+			variables,
+			coefficients,
+			&n_constraint_variables,
+			buffer_size,
+			&constant_offset,
+			&requiredsize,
+			true);
+	}
+
+	// Copy the relevant sections of the buffers to output vectors
+	std::vector<scip::Var*> output_variables;
+	std::vector<scip::real> output_coefficients;
+	output_variables.insert(output_variables.end(), variables, &variables[n_constraint_variables]);
+	output_coefficients.insert(output_coefficients.end(), coefficients, &coefficients[n_constraint_variables]);
+
+	// Obtain the left hand side
+	std::optional<scip::real> lhs;
+	scip::real lhs_value = SCIPconsGetLhs(scip, constraint, &success);
+	if (success && !SCIPisInfinity(scip, std::abs(lhs_value))) {
+		lhs = lhs_value - constant_offset;
+	} else
+		lhs = std::nullopt;
+
+	std::optional<scip::real> rhs;
+	scip::real rhs_value = SCIPconsGetRhs(scip, constraint, &success);
+	if (success && !SCIPisInfinity(scip, std::abs(rhs_value))) {
+		rhs = rhs_value - constant_offset;
+	} else
+		rhs = std::nullopt;
+
+	// Free the buffers
+	delete[] variables;
+	delete[] coefficients;
+
+	return std::make_tuple(output_variables, output_coefficients, lhs, rhs);
+}
+
+auto extract_constraints(scip::Model& model, bool normalize) -> std::tuple<utility::coo_matrix<value_type>, xmatrix> {
+	auto* const scip = model.get_scip_ptr();
+	auto const constraints = model.constraints();
+
+	std::size_t n_rows = 0;
+	auto n_cols = static_cast<std::size_t>(SCIPgetNVars(scip));
+
+	std::vector<value_type> values;
+	std::vector<std::size_t> column_indices;
+	std::vector<std::size_t> row_indices;
+	std::vector<value_type> biases;
```

Comment:
We can't know for `values`, `column_indices` and `row_indices` a priori because this depends on the transformation of constraints in terms of active variables only, which is as expensive as getting those coefficients in the first place.

For bias in principle we could have a loop and check which constraints have a lhs and rhs, and pre-allocate then. Do you think it's worth it? Alternatively, a rough upper bound is `2*nb_constraints`.

Reply:
Ok we decided to keep it as it is.